### PR TITLE
Use unsigned values for keeping track of gas 

### LINKED ISSUE #2574: Use unsigned values for keeping track of gas
**Issue summary**

We're currently using `i64` to keep track of gas limits and gas usage. Negative gas values do not make sense, though, and from nv18 we're moving to using `u64` instead.

Change the gas counters in `vm/message/src/lib.rs` to use `u64` and then propagate the changes through the rest of the code-base.

**Other information and links**

https://github.com/ChainSafe/forest/blob/5903083e7d71efc644a0d78134b554f99c86f6ab/vm/message/src/lib.rs#L34

### PR DESCRIPTION
**Pull request by tyshko5**
## Summary of changes

Changes introduced in this pull request:
- 

## Reference issue to close (if applicable)

Closes #2574

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: blockchain/chain/src/store/base_fee.rs [hunk 1]
```diff
@@ -10,40 +10,36 @@ use fvm_ipld_encoding::Cbor;
...
-    gas_limit_used: i64,
+    gas_limit_used: u64,
     no_of_blocks: usize,
     epoch: ChainEpoch,
     smoke_height: ChainEpoch,
 ) -> TokenAmount {
-    let mut delta: i64 = if epoch > smoke_height {
-        (gas_limit_used / no_of_blocks as i64) - BLOCK_GAS_TARGET
+    let mut delta: u64 = if epoch > smoke_height {
+        (gas_limit_used / no_of_blocks as u64) - BLOCK_GAS_TARGET
```
Comment (hanabi1224) [inline blockchain/chain/src/store/base_fee.rs:32]: u64 subtraction looks dangerous. The original logic does not guarantee delta to be non-negative

File: blockchain/chain/src/store/base_fee.rs [hunk 2]
```diff
@@ -10,40 +10,36 @@ use fvm_ipld_encoding::Cbor;
 use fvm_shared::{clock::ChainEpoch, BLOCK_GAS_LIMIT};
 
 /// Used in calculating the base fee change.
-pub const BLOCK_GAS_TARGET: i64 = BLOCK_GAS_LIMIT / 2;
+pub const BLOCK_GAS_TARGET: u64 = BLOCK_GAS_LIMIT as u64 / 2;
```
Comment (lemmih) [inline blockchain/chain/src/store/base_fee.rs:13]: You can use `BLOCK_GAS_LIMIT` from `fvm_shared3`

File: blockchain/chain/src/store/base_fee.rs (hunk 1 above)
Comment (tyshko5) [reply blockchain/chain/src/store/base_fee.rs]: Reverted

File: blockchain/chain/src/store/base_fee.rs (hunk 2 above)
Comment (tyshko5) [reply blockchain/chain/src/store/base_fee.rs]: Changed

File: vm/message/src/lib.rs [hunk 3]
```diff
@@ -63,8 +63,8 @@ impl Message for ShimMessage {
     fn params(&self) -> &RawBytes {
         &self.params
     }
-    fn gas_limit(&self) -> i64 {
-        self.gas_limit as i64
+    fn gas_limit(&self) -> u64 {
+        self.gas_limit
     }
     fn set_gas_limit(&mut self, token_amount: i64) {
```
Comment (lemmih) [inline vm/message/src/lib.rs:69]: This should be changed as well.
```suggestion
    fn set_gas_limit(&mut self, token_amount: u64) {
```

**Review by lemmih — CHANGES_REQUESTED** [review]
There are a few places where you can use `fvm_shared3` to your advantage.

File: blockchain/message_pool/src/msgpool/selection.rs [hunk 4]
```diff
@@ -157,12 +157,16 @@ where
...
             while next_chain < chains.len() {
                 let chain_key = chains.key_vec[next_chain];
                 next_chain += 1;
                 partitions[i].push(chain_key);
-                gas_limit -= chains.get(chain_key).unwrap().gas_limit;
+                let chain_gas_limit = chains.get(chain_key).unwrap().gas_limit;
+                if gas_limit < chain_gas_limit {
+                    break;
+                }
+                gas_limit -= chain_gas_limit;
```
Comment (hanabi1224) [inline blockchain/message_pool/src/msgpool/selection.rs:169]: Another scary subtraction

Comment (tyshko5) [reply blockchain/message_pool/src/msgpool/selection.rs:169]: 3 lines above we check that be can do this

File: vm/message/src/lib.rs (hunk 3 above)
Comment (tyshko5) [reply vm/message/src/lib.rs]: Fixed

**Review by hanabi1224 — APPROVED** [review]

**Review by lemmih — APPROVED** [review]

**Pull request merged**